src/ucp/core/ucp_worker.c

```diff
@@ -2385,6 +2385,100 @@ static void ucp_worker_set_max_am_header(ucp_worker_h worker)
                             ucs_min(max_am_header, UINT32_MAX) : 0ul;
 }
 
+static unsigned ucp_worker_progress_usage_tracker(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    ucs_time_t now;
+
+    if ((worker->usage_tracker.iter_count++ % UCP_WORKER_SLOW_CB_ITER_SKIP) !=
+        0) {
+        return 0;
+    }
+
+    now = ucs_get_time();
+    if (ucs_likely((now - worker->usage_tracker.last_round) <
+                   worker->context->config.ext.usage_tracker_interval)) {
+        return 0;
+    }
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    ucs_usage_tracker_progress(worker->usage_tracker.handle);
+    uct_worker_progress_unregister_safe(worker->uct,
+                                        &worker->usage_tracker.cb_id);
+    worker->usage_tracker.running = 1;
+
+    UCS_ASYNC_UNBLOCK(&worker->async);
+    return 1;
+}
+
+void ucp_worker_track_ep_usage(ucp_worker_h worker, ucp_ep_h ep)
+{
+    if (!worker->usage_tracker.running) {
+        return;
+    }
+
+    ucs_usage_tracker_touch_key(worker->usage_tracker.handle, ep);
+    worker->usage_tracker.samples_count++;
+
+    if ((worker->usage_tracker.samples_count %
+         UCS_USAGE_TRACKER_SAMPLES_COUNT_PER_RUN) == 0) {
+        worker->usage_tracker.running    = 0;
+        worker->usage_tracker.last_round = ucs_get_time();
+
+        uct_worker_progress_register_safe(worker->uct,
```

Comment:
maybe progress the usage tracker directly from here, instead of registering a progress callback?

Reply:
this code is fast path, shouldn't it bail out as soon as possible?